datasets/sharc/sharc.py

```diff
@@ -0,0 +1,126 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""ShARC: A Conversational Question Answering dataset focussing on question answering from texts containing rules."""
+
+from __future__ import absolute_import, division, print_function
+
+import json
+import os
+
+import datasets
+
+
+_CITATION = """\
+@misc{saeidi2018interpretation,
+      title={Interpretation of Natural Language Rules in Conversational Machine Reading},
+      author={Marzieh Saeidi and Max Bartolo and Patrick Lewis and Sameer Singh and Tim Rocktäschel and Mike Sheldon and Guillaume Bouchard and Sebastian Riedel},
+      year={2018},
+      eprint={1809.01494},
+      archivePrefix={arXiv},
+      primaryClass={cs.CL}
+}
+"""
+
+_DESCRIPTION = """\
+ShARC is a Conversational Question Answering dataset focussing on question answering from texts containing rules. \
+The goal is to answer questions by possibly asking follow-up questions first. It is assumed assume that the question is often underspecified, \
+in the sense that the question does not provide enough information to be answered directly. However, an agent can use the supporting rule text to \
+infer what needs to be asked in order to determine the final answer.
+"""
+
+_URL = "https://sharc-data.github.io/data/sharc1-official.zip"
+
+
+class Sharc(datasets.GeneratorBasedBuilder):
+    """ShARC: A Conversational Question Answering dataset focussing on question answering from texts containing rules."""
+
+    VERSION = datasets.Version("1.0.0")
+    BUILDER_CONFIGS = [
+        datasets.BuilderConfig(name="sharc", version=datasets.Version("1.0.0")),
+    ]
+
+    def _info(self):
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "id": datasets.Value("string"),
+                    "utterance_id": datasets.Value("string"),
+                    "source_url": datasets.Value("string"),
+                    "snippet": datasets.Value("string"),
+                    "question": datasets.Value("string"),
+                    "scenario": datasets.Value("string"),
+                    "history": [
+                        {"follow_up_question": datasets.Value("string"), "follow_up_answer": datasets.Value("string")}
+                    ],
+                    "evidence": [
+                        {"follow_up_question": datasets.Value("string"), "follow_up_answer": datasets.Value("string")}
+                    ],
+                    "answer": datasets.Value("string"),
+                    "negative_question": datasets.Value("bool_"),
+                    "negative_scenario": datasets.Value("bool_"),
```

Comment:
The dataset also provides ids of utterances where question or scenario is negative, so added two extra keys to indicate `negative_question` and `negative_scenario`